README.md

![wow](https://user-images.githubusercontent.com/15782257/71815946-bbf0e180-3089-11ea-9d00-10e05c308bea.png)

## Example

Take a look at the provided [example](https://github.com/kalaganov/outline_gradient_button/blob/master/example/main.dart) for how to use the OutlineGradientButton.

## How to use
![demo_button1](https://user-images.githubusercontent.com/15782257/71816114-41749180-308a-11ea-8673-dbd55cfdc940.png)
```dart
OutlineGradientButton(
  child: Text('WOW'),
  gradient: LinearGradient(
    colors: [Colors.purple, Colors.pink],
    begin: Alignment.topCenter,
    end: Alignment.bottomCenter,
  ),
  strokeWidth: 4,
),
```
--------
![demo_button2](https://user-images.githubusercontent.com/15782257/71816200-88628700-308a-11ea-8890-039ce04e5adf.png)
```dart
OutlineGradientButton(
  child: Text('TEXT'),
  gradient: LinearGradient(colors: [Colors.pink, Colors.purple]),
  strokeWidth: 4,
  radius: Radius.circular(8),
),
```
--------
![demo_button3](https://user-images.githubusercontent.com/15782257/71816269-bba51600-308a-11ea-876b-2577172b5a91.png)
```dart
OutlineGradientButton(
  child: Text('OMG'),
  gradient: LinearGradient(
    stops: [0, 0.5, 0.5, 1],
    colors: [Colors.green, Colors.green, Colors.red, Colors.red],
  ),
  strokeWidth: 4,
  corners: Corners(topLeft: Radius.elliptical(16, 14), bottomRight: Radius.circular(6)),
),
```
--------
![demo_button4](https://user-images.githubusercontent.com/15782257/71816322-ec854b00-308a-11ea-8444-f3cab9d58054.png)
```dart
OutlineGradientButton(
  child: SizedBox(
    width: 52,
    height: 52,
    child: Column(
      mainAxisSize: MainAxisSize.min,
      mainAxisAlignment: MainAxisAlignment.center,
      children: <Widget>[
        Icon(Icons.opacity, color: Colors.grey, size: 24),
        Text('TEXT', style: TextStyle(fontSize: 9)),
      ],
    ),
  ),
  gradient: LinearGradient(
    colors: [Colors.blue, Colors.black],
    begin: Alignment(-1, -1),
    end: Alignment(2, 2),
  ),
  strokeWidth: 4,
  padding: EdgeInsets.zero,
  radius: Radius.circular(26),
),
```
--------
![demo_button5](https://user-images.githubusercontent.com/15782257/71816465-59004a00-308b-11ea-82d6-74845de61835.png)
```dart
OutlineGradientButton(
  child: Text('Linear gradient', style: TextStyle(color: Colors.black, fontSize: 10)),
  gradient: LinearGradient(
    colors: List.generate(360, (h) => HSLColor.fromAHSL(1, h.toDouble(), 1, 0.5).toColor()),
  ),
  strokeWidth: 2,
  padding: EdgeInsets.symmetric(horizontal: 24, vertical: 12),
  radius: Radius.circular(8),
),
```
--------
![demo_button6](https://user-images.githubusercontent.com/15782257/71816560-af6d8880-308b-11ea-801b-d25d82526ae2.png)
```dart
OutlineGradientButton(
  child: Text('Sweep gradient', style: TextStyle(color: Colors.black, fontSize: 10)),
  gradient: SweepGradient(
    colors: List.generate(360, (h) => HSLColor.fromAHSL(1, h.toDouble(), 1, 0.5).toColor()),
  ),
  strokeWidth: 2,
  padding: EdgeInsets.symmetric(horizontal: 24, vertical: 12),
  radius: Radius.circular(8),
),
```
--------
![demo_button7](https://user-images.githubusercontent.com/15782257/71816617-e0e65400-308b-11ea-987c-d9ea80a21bda.png)
```dart
OutlineGradientButton(
  child: Text('With background color and elevation', style: TextStyle(color: Colors.white, fontSize: 12)),
  gradient: LinearGradient(colors: [Colors.greenAccent, Colors.yellow]),
  strokeWidth: 2,
  padding: EdgeInsets.symmetric(horizontal: 24, vertical: 12),
  corners: Corners(topRight: Radius.circular(16), bottomRight: Radius.circular(16)),
  backgroundColor: Colors.lightBlue,
  elevation: 4,
  inkWell: true,
  onTap: () => showSnack('onTap'),
  onDoubleTap: () => showSnack('onDoubleTap'),
  onLongPress: () => showSnack('onLongPress'),
),
```


CHANGELOG.md
## [2.0.1+1] - 11.03.2023

* fixed showing an example code on pub.dev

## [2.0.1] - 11.03.2023

* Increased minimum SDK version to 2.17.0
* minor fixes

## [2.0.0] - 26.03.2021

* Migrate to null safety.

## [1.0.0] - 01.06.2020

* Release version


example/body.dart
import 'package:flutter/material.dart';
import 'package:outline_gradient_button/outline_gradient_button.dart';

class Body extends StatelessWidget {
  const Body({super.key});

  @override
  Widget build(BuildContext context) {
    return Align(
      alignment: Alignment.topCenter,
      child: SingleChildScrollView(
        padding: EdgeInsets.all(48),
        child: Wrap(
          spacing: 24,
          runSpacing: 48,
          runAlignment: WrapAlignment.center,
          alignment: WrapAlignment.center,
          crossAxisAlignment: WrapCrossAlignment.center,
          children: [
            OutlineGradientButton(
              child: Text('WOW'),
              gradient: LinearGradient(
                colors: [Colors.purple, Colors.pink],
                begin: Alignment.topCenter,
                end: Alignment.bottomCenter,
              ),
              strokeWidth: 4,
            ),
            OutlineGradientButton(
              child: Text('TEXT'),
              gradient: LinearGradient(
                colors: [Colors.pink, Colors.purple],
              ),
              strokeWidth: 4,
              radius: Radius.circular(8),
            ),
            OutlineGradientButton(
              child: Text('OMG'),
              gradient: LinearGradient(
                stops: [0, 0.5, 0.5, 1],
                colors: [Colors.green, Colors.green, Colors.red, Colors.red],
              ),
              strokeWidth: 4,
              corners: Corners(
                topLeft: Radius.elliptical(16, 14),
                bottomRight: Radius.circular(6),
              ),
            ),
            OutlineGradientButton(
              child: SizedBox(
                width: 52,
                height: 52,
                child: Column(
                  mainAxisSize: MainAxisSize.min,
                  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                  children: <Widget>[
                    FlutterLogo(size: 16),
                    Text('TEXT', style: TextStyle(fontSize: 9)),
                  ],
                ),
              ),
              gradient: LinearGradient(
                colors: [Colors.blue, Colors.black],
                begin: Alignment(-1, -1),
                end: Alignment(2, 2),
              ),
              strokeWidth: 4,
              padding: EdgeInsets.zero,
              radius: Radius.circular(26),
            ),
            OutlineGradientButton(
              child: Text(
                'Linear gradient',
                style: TextStyle(color: Colors.black, fontSize: 10),
              ),
              gradient: LinearGradient(
                colors: List.generate(360, _generator),
              ),
              strokeWidth: 2,
              padding: EdgeInsets.symmetric(horizontal: 24, vertical: 12),
              radius: Radius.circular(8),
            ),
            OutlineGradientButton(
              child: Text(
                'Sweep gradient',
                style: TextStyle(color: Colors.black, fontSize: 10),
              ),
              gradient: SweepGradient(
                colors: List.generate(360, _generator),
              ),
              strokeWidth: 2,
              padding: EdgeInsets.symmetric(horizontal: 24, vertical: 12),
              radius: Radius.circular(8),
            ),
            OutlineGradientButton(
              child: Text(
                'With background color and elevation',
                style: TextStyle(color: Colors.white, fontSize: 12),
              ),
              gradient: LinearGradient(
                colors: [Colors.greenAccent, Colors.yellow],
              ),
              strokeWidth: 2,
              padding: EdgeInsets.symmetric(horizontal: 24, vertical: 12),
              corners: Corners(
                topRight: Radius.circular(16),
                bottomRight: Radius.circular(16),
              ),
              backgroundColor: Colors.lightBlue,
              elevation: 4,
              inkWell: true,
              onTap: () => _showPopup(context, 'onTap'),
              onDoubleTap: () => _showPopup(context, 'onDoubleTap'),
              onLongPress: () => _showPopup(context, 'onLongPress'),
            ),
          ],
        ),
      ),
    );
  }

  static void _showPopup(BuildContext context, String text) {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text(text)),
    );
  }

  static Color _generator(int h) {
    return HSLColor.fromAHSL(1, h.toDouble(), 1, 0.5).toColor();
  }
}


lib/outline_gradient_button.dart
library outline_gradient_button;

import 'package:flutter/material.dart';

/// OutlineGradientButton
class OutlineGradientButton extends StatelessWidget {
  final double strokeWidth;
  final Radius? radius;
  final Corners? corners;
  final Gradient gradient;
  final EdgeInsets padding;
  final Color backgroundColor;
  final double elevation;
  final bool inkWell;
  final GestureTapCallback? onTap;
  final GestureTapCallback? onDoubleTap;
  final GestureLongPressCallback? onLongPress;
  final GestureTapDownCallback? onTapDown;
  final GestureTapCancelCallback? onTapCancel;
  final ValueChanged<bool>? onHighlightChanged;
  final ValueChanged<bool>? onHover;
  final ValueChanged<bool>? onFocusChange;
  final BorderRadius _borderRadius;
  final Widget child;

  OutlineGradientButton({
    required this.strokeWidth,
    required this.gradient,
    this.corners,
    this.radius,
    this.padding = const EdgeInsets.all(8),
    this.backgroundColor = Colors.transparent,
    this.elevation = 0,
    this.inkWell = false,
    this.onTap,
    this.onDoubleTap,
    this.onLongPress,
    this.onTapDown,
    this.onTapCancel,
    this.onHighlightChanged,
    this.onHover,
    this.onFocusChange,
    required this.child,
    super.key,
  })  : assert(strokeWidth > 0),
        assert(padding.isNonNegative),
        assert(elevation >= 0),
        assert(radius == null || corners == null,
            'Cannot provide both a radius and corners.'),
        _borderRadius = corners != null
            ? _fromCorners(corners, strokeWidth)
            : _fromRadius(radius ?? Radius.zero, strokeWidth);

  @override
  Widget build(BuildContext context) {
    return Material(
      color: backgroundColor,
      elevation: elevation,
      borderRadius: _borderRadius,
      child: InkWell(
        borderRadius: _borderRadius,
        highlightColor:
            inkWell ? Theme.of(context).highlightColor : Colors.transparent,
        splashColor:
            inkWell ? Theme.of(context).splashColor : Colors.transparent,
        onTap: onTap,
        onLongPress: onLongPress,
        onDoubleTap: onDoubleTap,
        onTapDown: onTapDown,
        onTapCancel: onTapCancel,
        onHighlightChanged: onHighlightChanged,
        onHover: onHover,
        onFocusChange: onFocusChange,
        child: CustomPaint(
          painter: _Painter(gradient, radius, strokeWidth, corners),
          child: Padding(padding: padding, child: child),
        ),
      ),
    );
  }

  ///
  static BorderRadius _fromCorners(Corners corners, double strokeWidth) {
    return BorderRadius.only(
      topLeft: Radius.elliptical(
        corners.topLeft.x + strokeWidth,
        corners.topLeft.y + strokeWidth,
      ),
      topRight: Radius.elliptical(
        corners.topRight.x + strokeWidth,
        corners.topRight.y + strokeWidth,
      ),
      bottomLeft: Radius.elliptical(
        corners.bottomLeft.x + strokeWidth,
        corners.bottomLeft.y + strokeWidth,
      ),
      bottomRight: Radius.elliptical(
        corners.bottomRight.x + strokeWidth,
        corners.bottomRight.y + strokeWidth,
      ),
    );
  }

  ///
  static BorderRadius _fromRadius(Radius radius, double strokeWidth) {
    return BorderRadius.all(
      Radius.elliptical(
        radius.x + strokeWidth,
        radius.y + strokeWidth,
      ),
    );
  }
}

///
class _Painter extends CustomPainter {
  final Gradient gradient;
  final Radius? radius;
  final double strokeWidth;
  final Corners? corners;

  _Painter(this.gradient, this.radius, this.strokeWidth, this.corners);

  @override
  void paint(Canvas canvas, Size size) {
    final rect = _createRect(strokeWidth, size);
    final rRect = corners != null
        ? _createRRectFromRectAndCorners(rect, corners!)
        : RRect.fromRectAndRadius(rect, radius ?? Radius.zero);
    final paint = Paint()
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth
      ..shader = gradient.createShader(rect);
    canvas.drawRRect(rRect, paint);
  }

  ///
  static Rect _createRect(double strokeWidth, Size size) {
    return Rect.fromLTWH(
      strokeWidth / 2,
      strokeWidth / 2,
      size.width - strokeWidth,
      size.height - strokeWidth,
    );
  }

  ///
  static RRect _createRRectFromRectAndCorners(Rect rect, Corners corners) {
    return RRect.fromRectAndCorners(
      rect,
      topLeft: corners.topLeft,
      topRight: corners.topRight,
      bottomLeft: corners.bottomLeft,
      bottomRight: corners.bottomRight,
    );
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => oldDelegate != this;
}

///
class Corners {
  final Radius topLeft;
  final Radius topRight;
  final Radius bottomLeft;
  final Radius bottomRight;

  const Corners({
    this.topLeft = Radius.zero,
    this.topRight = Radius.zero,
    this.bottomLeft = Radius.zero,
    this.bottomRight = Radius.zero,
  });
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:outline_gradient_button/outline_gradient_button.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      theme: Theme.of(context).copyWith(useMaterial3: true),
      home: Scaffold(
        appBar: AppBar(
          title: Text('Outline gradient buttons demo'),
        ),
        backgroundColor: Colors.white,
        body: const Body(),
      ),
    );
  }
}

class Body extends StatelessWidget {
  const Body({super.key});

  @override
  Widget build(BuildContext context) {
    return Align(
      alignment: Alignment.topCenter,
      child: SingleChildScrollView(
        padding: EdgeInsets.all(48),
        child: Wrap(
          spacing: 24,
          runSpacing: 48,
          runAlignment: WrapAlignment.center,
          alignment: WrapAlignment.center,
          crossAxisAlignment: WrapCrossAlignment.center,
          children: [
            OutlineGradientButton(
              child: Text('WOW'),
              gradient: LinearGradient(
                colors: [Colors.purple, Colors.pink],
                begin: Alignment.topCenter,
                end: Alignment.bottomCenter,
              ),
              strokeWidth: 4,
            ),
            OutlineGradientButton(
              child: Text('TEXT'),
              gradient: LinearGradient(
                colors: [Colors.pink, Colors.purple],
              ),
              strokeWidth: 4,
              radius: Radius.circular(8),
            ),
            OutlineGradientButton(
              child: Text('OMG'),
              gradient: LinearGradient(
                stops: [0, 0.5, 0.5, 1],
                colors: [Colors.green, Colors.green, Colors.red, Colors.red],
              ),
              strokeWidth: 4,
              corners: Corners(
                topLeft: Radius.elliptical(16, 14),
                bottomRight: Radius.circular(6),
              ),
            ),
            OutlineGradientButton(
              child: SizedBox(
                width: 52,
                height: 52,
                child: Column(
                  mainAxisSize: MainAxisSize.min,
                  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                  children: <Widget>[
                    FlutterLogo(size: 16),
                    Text('TEXT', style: TextStyle(fontSize: 9)),
                  ],
                ),
              ),
              gradient: LinearGradient(
                colors: [Colors.blue, Colors.black],
                begin: Alignment(-1, -1),
                end: Alignment(2, 2),
              ),
              strokeWidth: 4,
              padding: EdgeInsets.zero,
              radius: Radius.circular(26),
            ),
            OutlineGradientButton(
              child: Text(
                'Linear gradient',
                style: TextStyle(color: Colors.black, fontSize: 10),
              ),
              gradient: LinearGradient(
                colors: List.generate(360, _generator),
              ),
              strokeWidth: 2,
              padding: EdgeInsets.symmetric(horizontal: 24, vertical: 12),
              radius: Radius.circular(8),
            ),
            OutlineGradientButton(
              child: Text(
                'Sweep gradient',
                style: TextStyle(color: Colors.black, fontSize: 10),
              ),
              gradient: SweepGradient(
                colors: List.generate(360, _generator),
              ),
              strokeWidth: 2,
              padding: EdgeInsets.symmetric(horizontal: 24, vertical: 12),
              radius: Radius.circular(8),
            ),
            OutlineGradientButton(
              child: Text(
                'With background color and elevation',
                style: TextStyle(color: Colors.white, fontSize: 12),
              ),
              gradient: LinearGradient(
                colors: [Colors.greenAccent, Colors.yellow],
              ),
              strokeWidth: 2,
              padding: EdgeInsets.symmetric(horizontal: 24, vertical: 12),
              corners: Corners(
                topRight: Radius.circular(16),
                bottomRight: Radius.circular(16),
              ),
              backgroundColor: Colors.lightBlue,
              elevation: 4,
              inkWell: true,
              onTap: () => _showPopup(context, 'onTap'),
              onDoubleTap: () => _showPopup(context, 'onDoubleTap'),
              onLongPress: () => _showPopup(context, 'onLongPress'),
            ),
          ],
        ),
      ),
    );
  }

  static void _showPopup(BuildContext context, String text) {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text(text)),
    );
  }

  static Color _generator(int h) {
    return HSLColor.fromAHSL(1, h.toDouble(), 1, 0.5).toColor();
  }
}
